File: nfr_odometry/src/nfr_odometry_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/transform_listener.h>
#include <tf2_ros/buffer.h>
#include <map>
#include <nlohmann/json.hpp>
#include <filesystem>
#include <fstream>
using namespace rclcpp;
using namespace std;
using Odometry = nav_msgs::msg::Odometry;
using TransformStamped = geometry_msgs::msg::TransformStamped;
using TransformBroadcaster = tf2_ros::TransformBroadcaster;
using TransformBuffer = tf2_ros::Buffer;
using TransformListener = tf2_ros::TransformListener;
using std::placeholders::_1;
class OdometryNode : public Node
{
private:
    shared_ptr<TransformBroadcaster> broadcaster;
    shared_ptr<TransformListener> listener;
    shared_ptr<TransformBuffer> buffer;
    Subscription<Odometry>::SharedPtr odometrySubscription;
public:
    OdometryNode() : Node("odometry_node")
    {
        initializeTransforms();
        string odometryTopic = declare_parameter("odom_topic", "odom");
        initializeOdometry(odometryTopic);
    }
    void initializeTransforms()
    {
        broadcaster = std::make_shared<TransformBroadcaster>(*this);
        buffer = std::make_shared<TransformBuffer>(get_clock());
        listener = std::make_shared<TransformListener>(*buffer);
    }
    void initializeOdometry(string odometryTopic)
    {
        odometrySubscription = create_subscription<Odometry>(odometryTopic, 10, bind(&OdometryNode::odometryCallback, this, _1));
        RCLCPP_INFO(get_logger(), "Started subscribing to odometry");
    }
    void odometryCallback(const Odometry& odometry)
    {
        TransformStamped transform;
        transform.header = odometry.header;
        transform.child_frame_id = odometry.child_frame_id;
        transform.transform.translation.x = odometry.pose.pose.position.x;
        transform.transform.translation.y = odometry.pose.pose.position.y;
        transform.transform.translation.z = 0;
        transform.transform.rotation = odometry.pose.pose.orientation;
        broadcaster->sendTransform(transform);
    }
};
#include <stdio.h>
#include <execinfo.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
void handler(int sig)
{
    void* array[10];
    size_t size;
    size = backtrace(array, 10);
    backtrace_symbols_fd(array, size, STDERR_FILENO);
    exit(1);
}
int main(int argc, char const *argv[])
{
    signal(SIGSEGV, handler);
    init(argc, argv);
    OdometryNode::SharedPtr node = make_shared<OdometryNode>();
    spin(node);
    shutdown();
    return 0;
}

File: nfr_odometry/nfr_odometry/nfr_odometry_node.py
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from nav_msgs.msg import Odometry
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster
class NFROdometryNode(Node):
    def __init__(self):
        super().__init__('nfr_odometry_node')
        self.odom_topic = self.declare_parameter('odom_topic', 'odom').value
        self.odom_subscription = self.create_subscription(Odometry, self.odom_topic, self.odometry_callback,
            qos_profile_sensor_data)
        self.tf_broadcaster = TransformBroadcaster(self)
        self.tf_timer = self.create_timer(0.05, self.publish_tf)
        self.odom_to_base_link = None
    def odometry_callback(self, odometry: Odometry):
        odometry_time = odometry.header.stamp.sec + odometry.header.stamp.nanosec / 1.0e+9
        current_stamp = self.get_clock().now().to_msg()
        current_time = current_stamp.sec + current_stamp.nanosec / 1.0e+9
        if abs(current_time - odometry_time) > 1:
            self.get_logger().warn('Recieved odometry message that is over one second late')
        if self.odom_to_base_link is None:
            self.odom_to_base_link = TransformStamped()
        self.odom_to_base_link.header = odometry.header
        self.odom_to_base_link.child_frame_id = odometry.child_frame_id
        self.odom_to_base_link.transform.rotation = odometry.pose.pose.orientation
        self.odom_to_base_link.transform.translation.x = odometry.pose.pose.position.x
        self.odom_to_base_link.transform.translation.y = odometry.pose.pose.position.y
        self.odom_to_base_link.transform.translation.z = odometry.pose.pose.position.z
    def publish_tf(self):
        if self.odom_to_base_link is not None:
            self.tf_broadcaster.sendTransform(self.odom_to_base_link)
def main(args=None):
    rclpy.init(args=args)
    odometry_node = NFROdometryNode()
    rclpy.spin(odometry_node)
    odometry_node.destroy_node()
    rclpy.shutdown()
if __name__ == '__main__':
    main()